### PR TITLE
Refactor of Dataset module and documentations and fixiing relevant errors

### PR DESCRIPTION
**Pull request by David-Morales-Norato**
This PR solves #148, #147 

Also improve the documentation of Datasets

### CODE REVIEW

**Comment by bemc22** [comment]
please fix the merge conflicts for review

**Comment by David-Morales-Norato** [comment]
> please fix the merge conflicts for review

Solved

File: colibri/data/sota_datasets.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import requests
+import zipfile
+import os 
+import numpy as np
+import PIL.Image as Image
+import torch
+
+class CaveDatasetHandler():
```
Comment (bemc22) [inline colibri/data/sota_datasets.py:8]: Should all subsequent datasets be labeled as `Handler`? I suggest naming them only by the dataset name, for example, `CaveDataset`.

File: colibri/data/sota_datasets.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        Load a sample from the CAVE dataset.
+        Args:
+            filename (str): The filename of the sample.
+        Returns:
+            dict: A dictionary containing the input and output data of the sample.
+        """
+        name = os.path.basename(filename).replace('_ms', '')
+
+        spectral_image = []
+        for i in range(1, 32):
```
Comment (bemc22) [inline colibri/data/sota_datasets.py:68]: Why does the loop run for 32 iterations? Is this a parameter of the dataset? I suggest assigning it to a variable for better readability.

File: colibri/data/sota_datasets.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            spectral_band_filename = os.path.join(filename, f'{name}_ms_{i:02d}.png')
+            spectral_band = np.array(Image.open(spectral_band_filename))
+            if len(spectral_band.shape)>2:
+                spectral_band = spectral_band[..., :3].mean(axis=-1)
```
Comment (bemc22) [inline colibri/data/sota_datasets.py:72]: I don't understand the purpose of this line.

Comment (David-Morales-Norato) [reply colibri/data/sota_datasets.py:72]: One image of the dataset has 4 channels (RGBA), the rest are RGB. the lines 71-72 handle this edge case

File: colibri/data/sota_datasets.py (hunk 2 above)
Comment (David-Morales-Norato) [reply colibri/data/sota_datasets.py:68]: solved

File: colibri/data/sota_datasets.py (hunk 1 above)
Comment (David-Morales-Norato) [reply colibri/data/sota_datasets.py:8]: solved

**Comment by David-Morales-Norato** [comment]
> I've made some minor changes directly and left other requests for code clarification.
> 
> For the rest, it's okay. Thanks for the PR.

solved

**Review by bemc22 — APPROVED** [review]

File: colibri/data/utils.py [hunk 4]
```diff
@@ -34,47 +60,9 @@ def load_builtin_dataset(name, path, **kwargs):
     dataset['input'] = (dataset['input'] / 255.).astype(np.float32)
     if dataset['input'].ndim != 4:
         dataset['input'] = dataset['input'].unsqueeze(1)
+    dataset['input'] = np.transpose(dataset['input'], (0, 3, 2, 1))
```
Comment (Enmartz) [inline colibri/data/utils.py:63]: The images are not oriented correctly. On a quick visualization of a sample, I got this:
image

 I suggest doing the following implementation instead:
`dataset['input'] = np.transpose(dataset['input'], (0, 3, 1, 2))`

If you apply this fix, be sure to update any places in your code where a similar permutation is being performed.

File: colibri/data/sota_datasets.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    The CAVE dataset is a database of multispectral images that were used to emulate the GAP camera. The images are of a wide variety of real-world materials and objects.
+
+    URL: https://www.cs.columbia.edu/CAVE/databases/multispectral/
+    """
+
+    def __init__(self, path: str, download: bool = True, url : str ='https://www.cs.columbia.edu/CAVE/databases/multispectral/zip/complete_ms_data.zip'):
+        self.url = url
+        self.tmp_name = 'cave_dataset'
+        self.path = os.path.join(path, self.tmp_name)
+        self.num_channels = 32
```
Comment (Enmartz) [inline colibri/data/sota_datasets.py:22]: The cave dataset has 31 spectral bands per sample. I understand that this is due to the indices given in the lesson on spectral images, but I suggest updating that part with +1 to compensate for this change.

File: colibri/data/datasets.py [hunk 6]
```diff
@@ -85,66 +84,58 @@ class CustomDataset(Dataset):
...
-        Arguments:
+    def __init__(self, name: str, path: str= "data", transform_dict:dict={}, **kwargs_builtin):
+        r"""
+        Args:
             name (string): Name of the dataset.
                 Current options are: ('cifar10', 'cifar100', 'mnist', 'fashion_mnist', 'cave').
             path (string): Path to directory with the dataset.
-            builtin_dict (dict,object): Dictionary with the parameters to load the builtin dataset.
             transform_dict (dict,object): Dictionary with the transformations to apply to the data.
+            kwargs_builtin (dict,object): Dictionary with the parameters to load the builtin datasets.
```
Comment (Enmartz) [inline colibri/data/datasets.py:94]: How can I determine the expected keys and values ​​for these dictionaries? I suggest specifying it here.

**Comment by Enmartz** [comment]
Thanks for your excellent work! I left some minor observations to be solved.

File: colibri/data/sota_datasets.py (hunk 5 above)
Comment (David-Morales-Norato) [reply colibri/data/sota_datasets.py:22]: solved

File: colibri/data/utils.py (hunk 4 above)
Comment (David-Morales-Norato) [reply colibri/data/utils.py:63]: solved

File: colibri/data/datasets.py (hunk 6 above)
Comment (David-Morales-Norato) [reply colibri/data/datasets.py:94]: solved

File: colibri/data/utils.py (hunk 4 above)
Comment (Enmartz) [reply colibri/data/utils.py]: When I run `demo_datasets.py`. I get this image:
image

Please adjust the visualization to the new permutation.

**Review by Enmartz — APPROVED** [review]

**Pull request merged**